client/driver/docker.go

```diff
@@ -355,35 +361,42 @@ func (d *DockerDriver) FSIsolation() cstructs.FSIsolation {
 	return cstructs.FSIsolationImage
 }
 
-func (d *DockerDriver) Prestart(ctx *ExecContext, task *structs.Task) error {
+func (d *DockerDriver) Prestart(ctx *ExecContext, task *structs.Task) (*CreatedResources, error) {
 	driverConfig, err := NewDockerDriverConfig(task, d.taskEnv)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
+	// Set state needed by Start()
+	d.driverConfig = driverConfig
+
 	// Initialize docker API clients
 	client, _, err := d.dockerClients()
 	if err != nil {
-		return fmt.Errorf("Failed to connect to docker daemon: %s", err)
+		return nil, fmt.Errorf("Failed to connect to docker daemon: %s", err)
 	}
 
+	// Ensure the image is available
 	if err := d.createImage(driverConfig, client, ctx.TaskDir); err != nil {
-		return err
+		return nil, err
 	}
 
-	image := driverConfig.ImageName
-	// Now that we have the image we can get the image id
-	dockerImage, err := client.InspectImage(image)
+	// Regardless of whether the image was downloaded already or not, store
+	// it as a created resource. Cleanup will soft fail if the image is
+	// still in use by another contianer.
+	dockerImage, err := client.InspectImage(driverConfig.ImageName)
 	if err != nil {
-		d.logger.Printf("[ERR] driver.docker: failed getting image id for %s: %s", image, err)
-		return fmt.Errorf("Failed to determine image id for `%s`: %s", image, err)
+		// When an error occurs we leak the image, but this should be
+		// extremely rare. There's no point in returning an error
+		// because the image won't be redownloaded as it now exists.
+		d.logger.Printf("[ERR] driver.docker: failed getting image id for %q: %v", driverConfig.ImageName, err)
+		return nil, nil
```

Comment:
This is a typo. The error should be returned here as I do require the image ID below!

Fixed.